src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,667 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// Represents a data structure in which each element additionally has a priority
+    /// associated with it.
+    /// </summary>
+    /// <typeparam name="TElement">The type of the element.</typeparam>
+    /// <typeparam name="TPriority">The type of the priority.</typeparam>
+    public class PriorityQueue<TElement, TPriority>
+    {
+        /// <summary>
+        /// Represents an implicit heap-ordered complete d-ary tree, stored as an array.
+        /// </summary>
+        private (TElement element, TPriority priority)[] _nodes;
+
+        private UnorderedItemsCollection? _unorderedItems;
+
+        /// <summary>
+        /// The number of nodes in the heap.
+        /// </summary>
+        private int _size;
+
+        /// <summary>
+        /// Used to keep enumerator in sync with the collection.
+        /// </summary>
+        private int _version;
```

Comment:
I see the API review comments in the issue say `It would not be snapshot, you can enumerate while modifying, just like any other collection`. 
If that's the case you can take the _version checks out entirely, and the code will be safe as the enumerator checks bounds every time it's called.

Reply:
> you can enumerate while modifying, just like any other collection

Those two statements aren't compatible.  Either there's a typo there or important context missing to the statement.

Reply:
@eiriktsarpalis do you know what was intended?

Reply:
We cannot guarantee sound enumeration while the collection is being modified, so the `_version` checks are there to invalidate older enumerators. This is consistent with other types in `System.Collections.Generic`.